clash-lib/src/Clash/Normalize/Transformations/Case.hs

```diff
@@ -0,0 +1,673 @@
+-- f :: Unsigned 4 -> Unsigned 4
+-- f x = case x of
+--   0 -> 3
+--   1 -> 2
+--   2 -> 1
+--   3 -> 0
+-- @
+--
+-- GHC generates Core that looks like:
+--
+-- @
+-- f = \(x :: Unsigned 4) -> case x == fromInteger 3 of
+--                             False -> case x == fromInteger 2 of
+--                               False -> case x == fromInteger 1 of
+--                                 False -> case x == fromInteger 0 of
+--                                   False -> error "incomplete case"
+--                                   True  -> fromInteger 3
+--                                 True -> fromInteger 2
+--                               True -> fromInteger 1
+--                             True -> fromInteger 0
+-- @
+--
+-- Which would result in a priority decoder circuit where a normal decoder
+-- circuit was desired.
+--
+-- This transformation transforms the above Core to the saner:
+--
+-- @
+-- f = \(x :: Unsigned 4) -> case x of
+--        _ -> error "incomplete case"
+--        0 -> fromInteger 3
+--        1 -> fromInteger 2
+--        2 -> fromInteger 1
+--        3 -> fromInteger 0
+-- @
+caseFlat :: HasCallStack => NormRewrite
+caseFlat _ e@(Case (collectEqArgs -> Just (scrut',_)) ty _) =
+  case collectFlat scrut' e of
+    Just alts' -> changed (Case scrut' ty (last alts' : init alts'))
+    Nothing -> return e
+
+caseFlat _ e = return e
+{-# SCC caseFlat #-}
+
+collectFlat :: Term -> Term -> Maybe [Alt]
+collectFlat scrut (Case (collectEqArgs -> Just (scrut', val)) _ty [lAlt,rAlt])
+  | scrut' == scrut
+  = case collectArgs val of
+      (Prim p,args') | isFromInt (primName p) ->
+        go (last args')
+      (Data dc,args') | nameOcc (dcName dc) == "GHC.Types.I#" ->
+        go (last args')
+      _ -> Nothing
+  where
+    go (Left (Literal i)) =
+      case (lAlt,rAlt) of
+        ((pl,el),(pr,er))
+          | isFalseDcPat pl || isTrueDcPat pr ->
+              case collectFlat scrut el of
+                Just alts' -> Just ((LitPat i, er) : alts')
+                Nothing -> Just [(LitPat i, er),(DefaultPat, el)]
+          | otherwise ->
+              case collectFlat scrut er of
+                Just alts' -> Just ((LitPat i, el) : alts')
+                Nothing -> Just [(LitPat i, el),(DefaultPat, er)]
+    go _ = Nothing
+
+    isFalseDcPat (DataPat p _ _)
+      = ((== "GHC.Types.False") . nameOcc . dcName) p
+    isFalseDcPat _ = False
+
+    isTrueDcPat (DataPat p _ _)
+      = ((== "GHC.Types.True") . nameOcc . dcName) p
+    isTrueDcPat _ = False
+
+collectFlat _ _ = Nothing
+{-# SCC collectFlat #-}
+
+collectEqArgs :: Term -> Maybe (Term,Term)
+collectEqArgs (collectArgsTicks -> (Prim pr, args, ticks))
+  | primName pr == "Clash.Sized.Internal.BitVector.eq#"
+  , [_, _, Left scrut, Left val] <- args
+  = Just (mkTicks scrut ticks, val)
+
+  | primName pr `elem` clashNumEqs
+  , [_, Left scrut, Left val] <- args
+  = Just (mkTicks scrut ticks, val)
+
+  | primName pr == "Clash.Transformations.eqInt"
```

Comment:
```show 'term```

Reply:
Not this one. The function doesn't really exist